### PR TITLE
Allow registration on matrix.org using a custom helper URL.

### PR DESCRIPTION
**Pull request by pixlwave**
Follow on to #3281, this PR adds the screen into the authentication flow.

### CODE REVIEW

File: ElementX/Sources/Application/AppSettings.swift [hunk 1]
```diff
@@ -159,6 +159,9 @@ final class AppSettings {
                                                                      staticRegistrations: oidcStaticRegistrations.mapKeys { $0.absoluteString },
                                                                      dynamicRegistrationsFile: .sessionsBaseDirectory.appending(path: "oidc/registrations.json"))
     
+    /// A temporary hack to allow registration on matrix.org until MAS is deployed.
+    let webRegistrationEnabled = false
```
Comment (pixlwave) [inline ElementX/Sources/Application/AppSettings.swift:163]: Note to self: Disabled for now until the required well-known is available on matrix.org

Comment (pixlwave) [reply ElementX/Sources/Application/AppSettings.swift:163]: Double note to self: If you enable this before merging, change the PR label.

File: ElementX/Sources/FlowCoordinators/AuthenticationFlowCoordinator.swift [hunk 2]
```diff
@@ -134,20 +137,20 @@ class AuthenticationFlowCoordinator: FlowCoordinatorProtocol {
...
+            showServerConfirmationScreen(authenticationFlow: flow)
         case .failure:
             stopLoading()
-            showServerSelectionScreen(isModallyPresented: false)
+            showServerSelectionScreen(authenticationFlow: flow, isModallyPresented: false)
         }
     }
     
-    private func showServerSelectionScreen(isModallyPresented: Bool) {
+    private func showServerSelectionScreen(authenticationFlow: AuthenticationFlow, isModallyPresented: Bool) {
```
Comment (stefanceriu) [inline ElementX/Sources/FlowCoordinators/AuthenticationFlowCoordinator.swift:153]: We need a state machine in here 😛

File: ElementX/Sources/FlowCoordinators/AuthenticationFlowCoordinator.swift [hunk 3]
```diff
@@ -166,13 +169,18 @@ class AuthenticationFlowCoordinator: FlowCoordinatorProtocol {
                     } else {
                         // We are here because the default server failed to respond.
                         if authenticationService.homeserver.value.loginMode == .password {
-                            // Add the password login screen directly to the flow, its fine.
-                            showLoginScreen()
+                            if authenticationFlow == .login {
+                                // Add the password login screen directly to the flow, its fine.
```
Comment (stefanceriu) [inline ElementX/Sources/FlowCoordinators/AuthenticationFlowCoordinator.swift:173]:
```suggestion
                                // Is it, really? 😁 
```

File: ElementX/Sources/Screens/Authentication/ServerConfirmationScreen/ServerConfirmationScreenViewModel.swift [hunk 4]
```diff
@@ -26,6 +26,7 @@ class ServerConfirmationScreenViewModel: ServerConfirmationScreenViewModelType,
             .sink { [weak self] homeserver in
                 guard let self else { return }
                 state.homeserverAddress = homeserver.address
+                state.homeserverSupportsRegistration = homeserver.loginMode == .oidc || homeserver.address == "matrix.org"
```
Comment (stefanceriu) [inline ElementX/Sources/Screens/Authentication/ServerConfirmationScreen/ServerConfirmationScreenViewModel.swift:29]: Outstanding! 😅

**Review by stefanceriu — APPROVED** [review]
lgtm, 🚀 it!

File: ElementX/Sources/FlowCoordinators/AuthenticationFlowCoordinator.swift (hunk 2 above)
Comment (pixlwave) [reply ElementX/Sources/FlowCoordinators/AuthenticationFlowCoordinator.swift:153]: Yep I agree, will refactor after the release.

File: ElementX/Sources/FlowCoordinators/AuthenticationFlowCoordinator.swift (hunk 3 above)
Comment (pixlwave) [reply ElementX/Sources/FlowCoordinators/AuthenticationFlowCoordinator.swift:173]: Lol. A follow-up refactor will be to move the request to the next screen (which matches Android) and then this failure will never happen.

File: ElementX/Sources/Application/AppSettings.swift (hunk 1 above)
Comment (pixlwave) [reply ElementX/Sources/Application/AppSettings.swift]: I've added an additional check for the URL so that this can be left as enabled.

**Pull request merged**